### PR TITLE
Make EqualsTester notice incorrect delegation

### LINKED ISSUE #877: Add delegated equality test to EqualsTester
Because python falls back to reference equality when `NotImplemented` is returned from `__eq__` methods, it is difficult to test the case where object A delegates its equality to object B by returning `NotImplemented` (instead of simply returning `False`, since A is not B whenever you do this).

A workaround for this would be a class that thinks it is equal to anything with the same hash code, or that it is equal to some specific instance. For example:

```python
class F:
    def __init__(self, other): self.other = other
    def __eq__(self, other): return True if other is self.other else NotImplemented
    def __hash__(self): return hash(self.other)
```

The idea is that, if the instance X under test returns False from `__eq__`, then `X == F(X)` would be `False` instead of `True`, and `X == F(None)` would still be `False`. This demonstrates that `X` is delegating to `F` (assuming you do both orderings) without needing to call `__eq__` directly (which I think we should avoid).

So: create a test that uses this class to detect classes that fail to return NotImplemented for values they don't understand, and add a test proving this actually works.

### PR DESCRIPTION
**Pull request by dabacon**
This extends EqualsTester to notice when someone incorrectly does not return NotImplemented in `__eq__` for objects of an unrecognized type.  Doing this can cause commutativity to fail when implementing other objects that should be equal to this object.  (It was worse in python 2, where depending on whether one inherited from object or not could cause this).

This also fixes places where Cirq code was doing this incorrectly.

Fixes #877

One final note is that the `eigen_gate_test` was changed because apparently Sympy does this incorrectly

```
>>> import sympy
>>> s = sympy.Symbol("a")
>>> class MyClass:
...   def __eq__(self, other):
...     return True
... 
>>> s.__eq__(MyClass())
False
```

### CODE REVIEW

**Comment by dabacon** [comment]
Cirq is secretly just a test of places where Sympy fails to be pythonic?

**Comment by balopat** [comment]
I just reran your checks as well on this PR to see the difference, couldn't repro the lint error yet on this branch. And now it passed. Very odd. XRef #3716.

**Comment by dabacon** [comment]
@Strilanc probably the best one to look at this sense he wrote the original definition of the problem (and there seems to be a slight difference in what he said, since we don't need a fallback to reference checking to get this work).

**Comment by dabacon** [comment]
@balopat this could use a triage to a reviewer

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
The comments in the test here are _incredibly_ helpful for parsing it, thanks!

One nit and this is good to go.

File: cirq/ops/eigen_gate_test.py [hunk 1]
```diff
@@ -105,8 +105,8 @@ def test_eq():
 
     eq.add_equality_group(CExpZinGate(2.5))
     eq.add_equality_group(CExpZinGate(2.25))
-    eq.make_equality_group(lambda: sympy.Symbol('a'))
-    eq.add_equality_group(sympy.Symbol('b'))
+    eq.make_equality_group(lambda: 'a')
+    eq.add_equality_group('b')
```
[the remark this answers is not in the document]
Comment (dabacon) [reply cirq/ops/eigen_gate_test.py:109]: Ah yeah, they are strange.  I guess it sort of makes sense to compare to some instance of another class sometimes (to check that you did an "isinstance check"), but I think equals tester already does this. Will remove.

**Review by 95-martin-orion — APPROVED** [review]

**Comment by dabacon** [comment]
tx @95-martin-orion !

**Pull request merged**